src/postOffice/LayerZeroPostman.sol

```diff
@@ -73,59 +63,38 @@ contract LayerZeroAdapter is ILayerZeroReceiver, ILayerZeroUserApplicationConfig
 
 	function lzReceive(
 		uint16 _srcChainId,
-		bytes memory _fromAddress,
+		bytes memory,
 		uint64, /*_nonce*/
 		bytes memory _payload
 	) external override {
 		// lzReceive can only be called by the LayerZero endpoint
 		if (msg.sender != address(endpoint)) revert Unauthorized();
 
-		// use assembly to extract the address from the bytes memory parameter
-		address fromAddress;
-		assembly {
-			fromAddress := mload(add(_fromAddress, 20))
-		}
-
 		// decode payload sent from source chain
-		(
-			uint256 _amount,
-			address _srcVaultAddress,
-			address _dstVaultAddress,
-			uint16 messageType
-		) = abi.decode(_payload, (uint256, address, address, uint16));
-
-		// TODO: Implement storage logic for differente message types.
-		// deposit has messageType === 1
-		// redeemRequest has messageType === 2
-		// redeem has messageType === 3
-
-		emit MessageReceived(
-			_srcChainId,
-			fromAddress,
-			_srcVaultAddress,
-			_amount,
-			_dstVaultAddress,
-			messageType
+		(Message memory _msg, address _dstVaultAddress, uint16 _messageType) = abi.decode(
+			_payload,
+			(Message, address, uint16)
 		);
+
+		emit MessageReceived(_msg.sender, _msg.value, _dstVaultAddress, _messageType, _srcChainId);
+
+		// send message to postOffice to be validated and processed
+		postOffice.writeMessage(_dstVaultAddress, _msg, messageType(_messageType));
 	}
 
-	function setChain(
-		uint256 _chainId,
-		address _adapter,
-		uint16 _lzChainId
-	) external onlyRole(MANAGER) {
-		chains[_chainId].adapter = _adapter;
-		chains[_chainId].lzChainId = _lzChainId;
+	// With this access control structure we need a way to vault set chain.
+	function setChain(uint16 _chainId, uint16 _lzChainId) external onlyRole(MANAGER) {
+		chains[_chainId] = _lzChainId;
 	}
 
 	function setConfig(
```

Comment:
This was changed in newer versions of the code. But basically, layerZero doesn't use the "standard" chainIDs. So, we need a mapping between EVM chainIDs and LayerZero chainIDs. We can't hardcode this directly because they can change those chainIDs (it happened recently). On the new version of this contract we set that mapping on the constructor, but the consequence for that is if LayerZero changes the chainIDs we need to deploy a new postman.

Reply:
The other setConfig, to be totally honest, is something very deep about LayerZero protocol that we don't fully understand yet. But basically we wrote this function because the interface needed. This is related if you want to change the standard behaviour of the postman, for example, if you want to use your own relayer and oracle instead of LayerZero's. Also, maybe we need that for upgradability because LayerZero is still changing things on their side.

Reply:
But documentation would help for sure. We are not proper documenting stuff right now because time and also for us still not clear if we settled for one implementation or other.

Reply:
👍